### PR TITLE
Multilocale library client now uses host compiler

### PR DESCRIPTION
**Pull request by dlongnecke-cray**
This PR adjusts multilocale libraries to compile the client library with the compiler specified by `CHPL_HOST_COMPILER`. The server is still compiled by the `CHPL_TARGET_COMPILER`. 

---

Testing:

- [ ] `ALL` on `linux64` with `CHPL_COMM=gasnet`, `HOST=gnu`, `TARGET=gnu`
- [x] `interop/C` on `darwin` with `CHPL_COMM=gasnet`, `HOST=clang`, `TARGET=clang`
- [ ] `interop/python` on `darwin` with `CHPL_COMM=gasnet`

Testing matrix for Crays, test file is `interop/C/multilocale/static/use_testing.ml-test.c` and `CHPL_COMM=ugni`:

- [ ] `HOST=gnu`, `TARGET=gnu`
- [ ] `HOST=gnu`, `TARGET=intel`
- [ ] `HOST=gnu`, `TARGET=cce`
- [ ] `HOST=intel`, `TARGET=gnu`
- [ ] `HOST=intel`, `TARGET=intel`
- [ ] `HOST=intel`, `TARGET=cce`
- [ ] `HOST=cce`, `TARGET=gnu`
- [ ] `HOST=cce`, `TARGET=intel`
- [ ] `HOST=cce`, `TARGET=cce`

### CODE REVIEW

File: runtime/etc/Makefile.launcher [hunk 1]
```diff
@@ -62,7 +62,7 @@ else
 # we need to make sure that we are generating position independent code,
 # IE with "-fPIC" on GCC.
 #
-	$(CC) $(GEN_CFLAGS) $(COMP_GEN_CFLAGS) $(SHARED_LIB_CFLAGS) \
+	$(CLIENT_CC) $(GEN_CFLAGS) $(COMP_GEN_CFLAGS) $(SHARED_LIB_CFLAGS) \
```
Comment (ronawho) [inline runtime/etc/Makefile.launcher]: What sets `CLIENT_CC` for non-mli builds?

File: runtime/etc/Makefile.mli-shared [hunk 2]
```diff
@@ -21,6 +21,15 @@
 
 include $(CHPL_MAKE_HOME)/runtime/etc/Makefile.include
 
+#
+# Taken from $CHPL_HOME/third-party/gmp/Makefile
```
Comment (ronawho) [inline runtime/etc/Makefile.mli-shared]: Don't think this comment is needed

File: runtime/etc/Makefile.mli-shared [hunk 3]
```diff
@@ -21,6 +21,15 @@
...
+#
+HOST_CC=$(shell CHPL_MAKE_HOME=$(CHPL_MAKE_HOME) CHPL_MAKE_HOST_TARGET=--host make -f $(CHPL_MAKE_HOME)/make/Makefile.base printQ-CC)
+TARG_CC=$(CC)
+
+CLIENT_CC=$(HOST_CC)
+SERVER_CC=$(TARG_CC)
```
Comment (ronawho) [inline runtime/etc/Makefile.mli-shared]: Seems a little odd to create client/host and server/targ aliases instead of just using one.

**Comment by ronawho** [comment]
Do you need to update sub_test to use the host compiler for mli builds?

File: runtime/etc/Makefile.launcher (hunk 1 above)
Comment (dlongnecke-cray) [reply runtime/etc/Makefile.launcher]: That code path should only run if the build is multilocale, else the existing code for executables is run.

```Makefile

ifneq ($(CHPL_MAKE_IS_MLI),1)
  $(CC) $(GEN_CFLAGS) $(COMP_GEN_CFLAGS) \
    -c -o $(TMPBINNAME)_launcher.o \
    -DLAUNCHER $(CHPL_LN_INCS) -I. $(LAUNCHER_SRC_NAME)
else
#
# If we are building launcher object files for a multi-locale library, then
# we need to make sure that we are generating position independent code,
# IE with "-fPIC" on GCC.
#
  $(CLIENT_CC) $(GEN_CFLAGS) $(COMP_GEN_CFLAGS) $(SHARED_LIB_CFLAGS) \
    -c -o $(TMPBINNAME)_launcher.o \
    -DLAUNCHER $(CHPL_LN_INCS) -I. $(LAUNCHER_SRC_NAME)
endif
```

File: runtime/etc/Makefile.mli-shared (hunk 3 above)
Comment (dlongnecke-cray) [reply runtime/etc/Makefile.mli-shared]: I just did it for legibility, but apparently it had the opposite effect.

**Comment by dlongnecke-cray** [comment]
Probably, I figure @lydia-duncan would know a bit more about that, and hopefully we could tackle that in a later patch.

**Comment by lydia-duncan** [comment]
I suspect the answer is yes.  I agree that it could be a later patch (potentially even after code/doc freeze, since iirc we don't bundle `start_test` into the release?  Though that would mean checking everything by hand instead of using `start_test`)

File: runtime/etc/Makefile.mli-shared (hunk 3 above)
Comment (lydia-duncan) [reply runtime/etc/Makefile.mli-shared]: I think a comment on the uses of whichever pair you keep would be helpful

**Comment by lydia-duncan** [comment]
In terms of how to test, our whitebox testing configuration is what tracks down the HOST-TARGET interaction.  It runs on a virtual machine but that's mostly so we don't overwhelm shared resources iirc.  An example configuration is:

```
CHPL_HOST_PLATFORM: linux64
CHPL_HOST_COMPILER: gnu
CHPL_HOST_ARCH: x86_64
CHPL_TARGET_PLATFORM: cray-xc *
CHPL_TARGET_COMPILER: cray-prgenv-intel
CHPL_TARGET_ARCH: x86_64
CHPL_TARGET_CPU: x86-64
CHPL_LOCALE_MODEL: flat
CHPL_COMM: none *
  CHPL_COMM_SUBSTRATE: none
  CHPL_GASNET_SEGMENT: none
CHPL_TASKS: qthreads
CHPL_LAUNCHER: none *
CHPL_TIMERS: generic
CHPL_UNWIND: none
CHPL_MEM: jemalloc
CHPL_ATOMICS: intrinsics
  CHPL_NETWORK_ATOMICS: none
CHPL_GMP: gmp
CHPL_HWLOC: hwloc
CHPL_REGEXP: re2
CHPL_LLVM: none
CHPL_AUX_FILESYS: none
CHPL_LIB_PIC: none
```

File: runtime/etc/Makefile.mli-shared [hunk 4]
```diff
@@ -22,14 +22,14 @@
 include $(CHPL_MAKE_HOME)/runtime/etc/Makefile.include
 
 #
-# Taken from $CHPL_HOME/third-party/gmp/Makefile
+# In multilocale library compilation, the client library is compiled with the
+# host compiler (CHPL_HOST_COMPILER), while the Chapel server (what is
```
Comment (lydia-duncan) [inline runtime/etc/Makefile.mli-shared]: "which is launched on the compute nodes"?

File: runtime/etc/Makefile.mli-shared [hunk 5]
```diff
@@ -44,11 +44,11 @@ all: buildUserObj buildServer buildClient runTagsCommand FORCE
 
 buildUserObj:
 #
-# Invoke the launcher script, but don't actually link up the launcher binary.
-# Do this by setting CHPL_MAKE_IS_MLI=1.
+# Compile loose source files in CHPLUSEROBJ. Use the TARG_CC for now, but
+# later we may need to roll these into both client and server.
```
Comment (lydia-duncan) [inline runtime/etc/Makefile.mli-shared]: I would remove everything after the comma and replace it with "as the user code will be run by the server" or something to that effect

File: runtime/etc/Makefile.mli-static [hunk 6]
```diff
@@ -46,11 +46,11 @@ all: buildUserObj buildServer buildClient runTagsCommand FORCE
 
 buildUserObj:
 #
-# Compile loose source files in CHPLUSEROBJ. Use the SERVER_CC for now, but
+# Compile loose source files in CHPLUSEROBJ. Use the TARG_CC for now, but
 # later we may need to roll these into both client and server.
```
Comment (lydia-duncan) [inline runtime/etc/Makefile.mli-static]: Ditto here, I guess

**Comment by dlongnecke-cray** [comment]
I'm going to close this for now, as additional work is required to solve linking errors before this code can be merged.

**Pull request closed without merging**